op-node/rollup/derive/batch_queue.go

```diff
@@ -4,205 +4,301 @@ import (
+		return fmt.Errorf("cannot add batch with timestamp %d, no origin was prepared", batch.Timestamp)
 	}
-	if uint64(len(bq.inputs)) > bq.config.SeqWindowSize {
-		return nil, fmt.Errorf("unexpectedly buffered more L1 inputs than sequencing window: %d", len(bq.inputs))
+
+	data := BatchWithL1InclusionBlock{
+		L1InclusionBlock: bq.progress.Origin,
+		Batch:            batch,
 	}
-	l1Origin := bq.inputs[0].Origin
-	nextL1Block := bq.inputs[1].Origin
+	batches, ok := bq.batchesByTimestamp[batch.Timestamp]
+	// Filter complete duplicates. This step is not strictly needed as we always append, but it is nice to avoid lots
+	// of spam. If we enforced a tighter relationship between block number & timestamp we could limit to a single
+	// blocknumber & epoch pair.
+	// TODO: Just log instead of error?
+	// TODO: am getting duplicate batches here, not sure why
+	if ok {
+		for _, b := range batches {
+			if b.Batch.Timestamp == batch.Timestamp && b.Batch.EpochNum == batch.EpochNum {
+				bq.log.Warn("duplicate batch", "batch_epoch", batch.EpochNum, "batch_timestamp", batch.Timestamp, "txs", len(batch.Transactions))
+				return nil
+			}
+		}
+	} else {
+		bq.log.Warn("First seen batch", "batch_epoch", batch.EpochNum, "batch_timestamp", batch.Timestamp, "txs", len(batch.Transactions))
 
-	fetchCtx, cancel := context.WithTimeout(ctx, 20*time.Second)
-	defer cancel()
-	l1Info, _, receipts, err := bq.dl.Fetch(fetchCtx, l1Origin.Hash)
-	if err != nil {
-		bq.log.Error("failed to fetch L1 block info", "l1Origin", l1Origin, "err", err)
-		return nil, nil
 	}
+	// May have duplicate block numbers or individual fields, but have limited complete duplicates
+	bq.batchesByTimestamp[batch.Timestamp] = append(batches, &data)
+	return nil
+}
 
-	deposits, errs := DeriveDeposits(receipts, bq.config.DepositContractAddress)
-	for _, err := range errs {
-		bq.log.Error("Failed to derive a deposit", "l1OriginHash", l1Origin.Hash, "err", err)
+// validExtension determines if a batch follows the previous attributes
+func validExtension(cfg *rollup.Config, batch *BatchWithL1InclusionBlock, prevTime, prevEpoch uint64) bool {
+	if batch.Batch.Timestamp != prevTime+cfg.BlockTime {
+		return false
 	}
-	if len(errs) != 0 {
-		return nil, fmt.Errorf("failed to derive some deposits: %v", errs)
+	if batch.Batch.EpochNum != rollup.Epoch(prevEpoch) && batch.Batch.EpochNum != rollup.Epoch(prevEpoch+1) {
+		return false
 	}
+	// TODO: Also check EpochHash
 
-	minL2Time := uint64(lastL2Timestamp) + bq.config.BlockTime
-	maxL2Time := l1Origin.Time + bq.config.MaxSequencerDrift
-	if minL2Time+bq.config.BlockTime > maxL2Time {
-		maxL2Time = minL2Time + bq.config.BlockTime
+	// TODO (VERY IMPORTANT): Filter this batch out if the origin of the batch is too far past the epoch
+	// TODO (ALSO VERY IMPT): Get this equality check correct
+	// TODO: Also do this check when ingesting batches.
+	if uint64(batch.Batch.EpochNum) >= batch.L1InclusionBlock.Number+cfg.SeqWindowSize {
+		return false
```

Comment:
The batch epoch num shouldn't be equal or higher than the inclusion block to begin with, since it's referencing that block by hash.
Do you mean:
```
epochNum := uint64(batch.Batch.EpochNum)
// note: overflow safe, since epochNum has been validated above to match current or next epoch. So it does not have an absurdly high value.
if epochNum+cfg.SeqWindowSize < batch.L1InclusionBlock.Number {
  return false
}
```

Reply:
yea, figured this out. I like that comparison